### DISCUSSION TITLE
Using the WebSpeech API to remove the Whisper model

### DISCUSSION THREAD

**Discussion by ndrean** [dwyl/image-classifier#103] [General]
Instead of using the "Whisper" model to transcript an audio into a text, we can use the native [Web Speech API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Speech_API/Using_the_Web_Speech_API).

> Note the compatibility status: not with Firefox .

This can save 1Gb.

You can try it with this script ([source MDN](https://github.com/mdn/dom-examples/blob/main/web-speech-api/phrase-matcher/script.js))

```js
// recognition.js
window.SpeechRecognition =
  Window.SpeechRecognition || window.webkitSpeechRecognition;

const recognition = new SpeechRecognition();
recognition.continuous = true;

recognition.onresult = (e) => {
  if (e.results[0].isFinal) {
    const p = document.createElement("p");
    p.textContent = Array.from(e.results)
      .map((result) => result[0])
      .map((res) => res.transcript)
      .join(", ");

    document.querySelector("#phrase").append(p);
  }
};

document.querySelector("#start").addEventListener("click", () => recognition.start());
document.querySelector("#stop").addEventListener("click", () => recognition.stop());
```

and serve this "index.html" file:

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Speech</title>
  </head>
  <body>
    <button id="start" type="button">Start</button>
    <button id="stop" type="button">Stop</button>
    <div id="phrase"></div>
    <script type="module" src="/recognition.js"></script>
  </body>
</html>
```

This could be used (an improved version) via a `hook` and push the result to the server to compute an embedding.

**Comment by LuchoTurtle** [comment]
A fine alternative 🤔 it'd be great to see if there was a way to compare the results between the two. `whisper` probably has this comparison

**Comment by ndrean** [comment]
All I can say is that the web speech seems pretty fast and accurate, much faster that the ML process. It seems to work only for the english language (even if  change "lang=es") whilst OpenAI Whisper can deal automatically with several languages.
I did not find any comparison. 

[Some explanations on how this works](https://developer.mozilla.org/en-US/docs/Web/API/Web_Speech_API/Using_the_Web_Speech_API#speech_recognition): this does not work offline.

I found another answer here:
<https://stackoverflow.com/questions/74945259/does-the-web-speech-api-in-chrome-and-edge-use-an-offsite-server-for-stt>
